hudi-common/src/test/java/org/apache/hudi/common/table/read/TestSchemaHandler.java

```diff
@@ -0,0 +1,464 @@
+                      boolean mergeUseRecordPosition,
+                      boolean supportsParquetRowIndex) {
+    HoodieReaderContext<String> readerContext = new MockReaderContext(supportsParquetRowIndex);
+    readerContext.setHasLogFiles(true);
+    readerContext.setHasBootstrapBaseFile(false);
+    //has no effect on schema unless we support position based merging
+    readerContext.setShouldMergeUseRecordPosition(mergeUseRecordPosition);
+    HoodieTableConfig hoodieTableConfig = mock(HoodieTableConfig.class);
+    when(hoodieTableConfig.populateMetaFields()).thenReturn(Boolean.TRUE);
+    when(hoodieTableConfig.getRecordMergeMode()).thenReturn(mergeMode);
+    if (hasPrecombine) {
+      when(hoodieTableConfig.getPreCombineField()).thenReturn("timestamp");
+    } else {
+      when(hoodieTableConfig.getPreCombineField()).thenReturn(null);
+    }
+    if (mergeMode == CUSTOM) {
+      Option<HoodieRecordMerger> merger = Option.of(new MockMerger(isProjectionCompatible, new String[]{"begin_lat", "begin_lon", "_hoodie_record_key", "timestamp"}));
+      readerContext.setRecordMerger(merger);
+    }
+    Schema requestedSchema = DATA_SCHEMA;
+    HoodieFileGroupReaderSchemaHandler schemaHandler = supportsParquetRowIndex
+        ? new HoodiePositionBasedSchemaHandler(readerContext, DATA_SCHEMA, requestedSchema,
+        Option.empty(), hoodieTableConfig, new TypedProperties())
+        : new HoodieFileGroupReaderSchemaHandler(readerContext, DATA_SCHEMA, requestedSchema,
+        Option.empty(), hoodieTableConfig, new TypedProperties());
+    Schema expectedRequiredFullSchema = supportsParquetRowIndex && mergeUseRecordPosition
+        ? HoodiePositionBasedSchemaHandler.addPositionalMergeCol(requestedSchema)
+        : requestedSchema;
+    assertEquals(expectedRequiredFullSchema, schemaHandler.getRequiredSchema());
+
+    //read subset of columns
+    requestedSchema = DATA_COLS_ONLY_SCHEMA;
+    schemaHandler = supportsParquetRowIndex
+        ? new HoodiePositionBasedSchemaHandler(readerContext, DATA_SCHEMA, requestedSchema,
+        Option.empty(), hoodieTableConfig, new TypedProperties())
+        : new HoodieFileGroupReaderSchemaHandler(readerContext, DATA_SCHEMA, requestedSchema,
+        Option.empty(), hoodieTableConfig, new TypedProperties());
+    Schema expectedRequiredSchema;
+    if (mergeMode == EVENT_TIME_ORDERING && hasPrecombine) {
+      expectedRequiredSchema = generateProjectionSchema("begin_lat", "tip_history", "rider", "_hoodie_record_key", "timestamp");
+    } else if (mergeMode == EVENT_TIME_ORDERING || mergeMode == COMMIT_TIME_ORDERING) {
+      expectedRequiredSchema = generateProjectionSchema("begin_lat", "tip_history", "rider", "_hoodie_record_key");
+    } else if (mergeMode == CUSTOM && isProjectionCompatible) {
+      expectedRequiredSchema = generateProjectionSchema("begin_lat", "tip_history", "rider", "begin_lon", "_hoodie_record_key", "timestamp");
+    } else {
+      expectedRequiredSchema = DATA_SCHEMA;
```

Comment:
don't we need to add meta fields at the beginning in addition to actual data cols? 
for CUSTOM merge mode and if projection is not compatible

Reply:
```
protected static final Schema DATA_SCHEMA = HoodieAvroUtils.addMetadataFields(HoodieTestDataGenerator.AVRO_SCHEMA);
```